cmake_unofficial/CMakeLists.txt

```diff
@@ -1,40 +1,108 @@
-cmake_minimum_required(VERSION 2.6)
-cmake_policy(VERSION 2.6)
+# To the extent possible under law, the author(s) have dedicated all
+# copyright and related and neighboring rights to this software to
+# the public domain worldwide. This software is distributed without
+# any warranty.
+#
+# For details, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 
-project(xxhash)
+set(XXHASH_TOP_SOURCE_DIR "${CMAKE_CURRENT_SOURCE_DIR}/..")
 
-set(XXHASH_LIB_VERSION "0.42.0")
-set(XXHASH_LIB_SOVERSION "0")
+# Parse version information
+file(STRINGS "${XXHASH_TOP_SOURCE_DIR}/xxhash.h" XXHASH_VERSION_MAJOR REGEX "^#define XXH_VERSION_MAJOR +([0-9]+) *$")
+string(REGEX REPLACE "^#define XXH_VERSION_MAJOR +([0-9]+) *$" "\\1" XXHASH_VERSION_MAJOR "${XXHASH_VERSION_MAJOR}")
+file(STRINGS "${XXHASH_TOP_SOURCE_DIR}/xxhash.h" XXHASH_VERSION_MINOR REGEX "^#define XXH_VERSION_MINOR +([0-9]+) *$")
+string(REGEX REPLACE "^#define XXH_VERSION_MINOR +([0-9]+) *$" "\\1" XXHASH_VERSION_MINOR "${XXHASH_VERSION_MINOR}")
+file(STRINGS "${XXHASH_TOP_SOURCE_DIR}/xxhash.h" XXHASH_VERSION_RELEASE REGEX "^#define XXH_VERSION_RELEASE +([0-9]+) *$")
+string(REGEX REPLACE "^#define XXH_VERSION_RELEASE +([0-9]+) *$" "\\1" XXHASH_VERSION_RELEASE "${XXHASH_VERSION_RELEASE}")
+set(XXHASH_VERSION_STRING "${XXHASH_VERSION_MAJOR}.${XXHASH_VERSION_MINOR}.${XXHASH_VERSION_RELEASE}")
+mark_as_advanced(XXHASH_VERSION_STRING XXHASH_VERSION_MAJOR XXHASH_VERSION_MINOR XXHASH_VERSION_RELEASE)
 
-option(BUILD_XXHSUM "Build the xxhsum binary" ON)
+if("${CMAKE_VERSION}" VERSION_LESS "3.0")
+  project(XXHASH C)
+else()
+  cmake_policy (SET CMP0048 NEW)
+  project(XXHASH
+    VERSION ${XXHASH_VERSION_STRING}
+    LANGUAGES C)
+endif()
 
-# Make CMake's RPATH handling not be insane. This suff has cmake set rpaths appropriately for
-# where things end up in the install tree. For some reason that's not the default:
-# https://cmake.org/Wiki/CMake_RPATH_handling
-SET(CMAKE_SKIP_BUILD_RPATH FALSE)
-SET(CMAKE_BUILD_WITH_INSTALL_RPATH FALSE)
+cmake_minimum_required (VERSION 2.8.6)
 
-# Where we search for shared libraries
-SET(CMAKE_INSTALL_RPATH "./lib")
+# If XXHASH is being bundled in another project, we don't want to
+# install anything.  However, we want to let people override this, so
+# we'll use the XXHASH_BUNDLED_MODE variable to let them do that; just
+# set it to OFF in your project before you add_subdirectory(xxhash/contrib/cmake_unofficial).
```

Comment:
Totally agree about it being a "hack"… it's kinda like executing a command from a C program instead of calling into a library; it has its place, but if you have a choice a library is generally a better way to go.

ExternalProject basically isolates the child project, and in order to carry things over you have to interact with its build system's interface.  If you apply something to a parent project (set compiler or linker flags, choose a compiler, add a define, etc.) it doesn't get carried through to the child.

Things also don't go the other way… you can't interact with targets automatically, you need to figure out where files were generated and recreate targets in the parent project, including paths, any required flags (i.e., whatever `pkg-config --cflags child` would spit out), etc.

If that's how someone wants to interact with xxHash's build system, that's still an option.  Bundled mode is basically just an additional option that provides more control and easier integration.

That said, I agree it's a bit strange.  I don't see it all that often in the wild, but I think it's a much nicer way to go.  Since it's not hurting anything, I don't see a reason not to have it.